tensordict/tensordict.py

```diff
@@ -4802,38 +4808,87 @@ def to(tensor):
             result.batch_size = batch_size
         return result
 
-    def where(self, condition, other, *, out=None):
-        if out is None:
-            if _is_tensor_collection(other.__class__):
-
-                def func(tensor, _other):
-                    return torch.where(
-                        expand_as_right(condition, tensor), tensor, _other
-                    )
+    def where(self, condition, other, *, out=None, pad=None):
+        if _is_tensor_collection(other.__class__):
 
-                return self._fast_apply(func, other)
-            else:
+            def func(tensor, _other, key):
+                if tensor is None:
+                    if pad is not None:
+                        tensor = _other
+                        _other = pad
+                    else:
+                        raise KeyError(
+                            f"Key {key} not found and no pad value provided."
+                        )
+                    cond = expand_as_right(~condition, tensor)
+                elif _other is None:
+                    if pad is not None:
+                        _other = pad
+                    else:
+                        raise KeyError(
+                            f"Key {key} not found and no pad value provided."
```

Comment:
Before this PR was an error thrown for these cases? What was the treatment for missing keys?

Reply:
it was a bug in the sense that `td0.where(cond, td1)` and `td1.where(~cond, td0)` where not returning the same thing. Only the keys of the first (self) were scanned. Keys in `self` but not `other` would raise an exception, but keys in `other` but not `self` where just ignored. This is super bad IMO